### PR TITLE
mempool: improve dump; fix buffer accounting bugs

### PR DESCRIPTION
**Pull request by liewegas**

### CODE REVIEW

**Comment by tchaikov** [comment]
retest this please. see #15406

**Comment by liewegas** [comment]
retest this please

File: src/common/buffer.cc [hunk 1]
```diff
@@ -1768,7 +1768,7 @@ static std::atomic_flag buffer_debug_lock = ATOMIC_FLAG_INIT;
   {
     if (append_buffer.unused_tail_length() < prealloc) {
       append_buffer = buffer::create(prealloc);
-      if (_mempool) {
+      if (_mempool >= 0) {
```
Comment (tchaikov) [inline src/common/buffer.cc]: yeah, it's `-1` by default. and the mem pool index starts at 0. they are enum whose first element is not assigned with a number.

File: src/test/test_mempool.cc [hunk 2]
```diff
@@ -277,16 +289,37 @@ TEST(mempool, bufferlist)
...
+  size_t items_before = mempool::buffer_anon::allocated_items();
+  size_t bytes_before = mempool::buffer_anon::allocated_bytes();
+  bl.append("fooo");
+  ASSERT_EQ(items_before + 1, mempool::buffer_anon::allocated_items());
+  ASSERT_LT(bytes_before, mempool::buffer_anon::allocated_bytes());
+
+  // move existing bl
+  bl.reassign_to_mempool(mempool::mempool_osd);
+  ASSERT_EQ(items_before, mempool::buffer_anon::allocated_items());
+  ASSERT_EQ(bytes_before, mempool::buffer_anon::allocated_bytes());
```
Comment (tchaikov) [inline src/test/test_mempool.cc]: nit, could check if `mempool::mempool_osd` gets the item and bytes number.

**Comment by wjwithagen** [comment]
@liewegas 
Something in this PR fixes a crash I'm having (on FreeBSD) since the previous changes to the buffer/mempool code, likely #15352.
So I'm all in favour for this PR. :)

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-06-02_08:32:01-rados-wip-sage-testing-distro-basic-smithi/

**Pull request merged**